pkg/gitops/filter.go

```diff
@@ -0,0 +1,90 @@
+package gitops
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/object"
+)
+
+func (cr *CustomRepo) FilterCommits(author string, since time.Time, until time.Time, resource string, namespace string, name string) ([]object.Commit, error) {
+	var logopts git.LogOptions
+	var filteredCommits []object.Commit
+
+	cr.Mutex.Lock()
+	defer cr.Mutex.Unlock()
+	ref, err := cr.Repo.Head()
+	if err != nil {
+		return filteredCommits, fmt.Errorf("could not get ref head from repository: %w", err)
+	}
+
+	logopts.From = ref.Hash()
+	if !since.IsZero() {
+		logopts.Since = &since
+	}
+	if !until.IsZero() {
+		logopts.Until = &until
+	}
+	var resources []string
+	if resource != "" && namespace == "" && name != "" {
+		return filteredCommits, fmt.Errorf("Error (FilterCommits): cannot provide a resource and a name without a namespace")
+	} else if resource == "" && namespace != "" {
+		resources = []string{"k8s-policies", "antrea-policies", "antrea-cluster-policies", "antrea-tiers"}
+	} else {
+		resources = []string{resource}
+	}
+	for _, r := range resources {
+		setPathFilter(r, namespace, name, &logopts)
+		tempCommits, err := cr.filter(&logopts, func(commit *object.Commit) bool {
+			if author == "" || commit.Author.Name == author {
+				return true
+			}
+			return false
+		})
+		if err != nil {
+			return filteredCommits, err
+		}
+		filteredCommits = append(filteredCommits, tempCommits...)
+	}
+	return filteredCommits, err
+}
+
+type customFilterFn func(*object.Commit) bool
+
+func (cr *CustomRepo) filter(logopts *git.LogOptions, customFilters ...customFilterFn) ([]object.Commit, error) {
+	var filteredCommits []object.Commit
+	cIter, err := cr.Repo.Log(logopts)
+	if err != nil {
+		return filteredCommits, fmt.Errorf("could not get logs from repo: %w", err)
+	}
+
+	err = cIter.ForEach(func(c *object.Commit) error {
+		passed := true
+		for _, filterFn := range customFilters {
+			if !filterFn(c) {
+				passed = false
+				break
+			}
+		}
+		if passed {
+			filteredCommits = append(filteredCommits, *c)
+		}
+		return nil
+	})
+	return filteredCommits, err
+}
+
+func setPathFilter(resource string, namespace string, name string, logopts *git.LogOptions) {
+	fpath := filepath.Join(resource, namespace, name)
+	if fpath != "" {
+		logopts.PathFilter = func(path string) bool {
+			if strings.Contains(path, fpath) {
+				return true
+			}
+			return false
+		}
+	}
```

Comment:
just want to make sure that this works well for all cases, e.g. if `namespace` is empty, I think we get `<resource>/<name>` as the path. Is the filtering correct in that case (I don't think so)?

Reply:
That's the only case it doesn't work (I even throw an error) because there might be any amount of user defined namespaces, so I don't know how I would check them all.

Reply:
`PathFilter` is an arbitrary function so you can use a regex in the implementation, as they mention in the go-git documentation: https://pkg.go.dev/github.com/go-git/go-git/v5#LogOptions